File: src/main/clojure/nico/ui/alert_panel.clj
;; -*- coding: utf-8-unix -*-
(ns #^{:author "sgr" :doc "Alert panel."}
  nico.ui.alert-panel
  (:require [clojure.tools.logging :as log]
            [nico.prefs :as p]
            [nico.ui.util :as uu]
            [net-utils :as n]
            [str-utils :as su]
            [time-utils :as tu])
  (:import [java.awt Color Component Container Cursor Dimension FlowLayout Font GraphicsEnvironment
            GridBagLayout GridBagConstraints Insets]
           [java.awt.event ActionListener ComponentAdapter MouseEvent MouseListener MouseMotionListener WindowEvent]
           [java.awt.geom RoundRectangle2D$Float]
           [java.net URL]
           [java.util.concurrent TimeUnit]
           [javax.swing BorderFactory ImageIcon JButton JDialog JLabel JPanel JTextArea SpringLayout]
           [javax.swing.text.html HTMLEditorKit]))

(def OPACITY (float 0.9))
(def ^{:private true} ^Dimension ASIZE (Dimension. 220 130))

(defn dlg-width  [] (.width  ASIZE))
(defn dlg-height [] (.height ASIZE))
(defn get-shape  [] (RoundRectangle2D$Float. 0 0 (.width ASIZE) (.height ASIZE) 20 20))

(def ^{:private true} ^Cursor LINK-CURSOR (.getLinkCursor (HTMLEditorKit.)))
(def ^{:private true} ^ImageIcon CLOSE-ICON (ImageIcon. (clojure.java.io/resource "closebtn.png")))
(def ^{:private true} ^Dimension CLOSE-ICON-SIZE (Dimension. (.getIconWidth CLOSE-ICON) (.getIconHeight CLOSE-ICON)))
(def ^{:private true} ^Dimension TITLE-PANEL-SIZE (Dimension. 210 18))
(def ^{:private true} ^Color MONLY-BGCOLOR (Color. 165 204 255))
(def ^{:private true} ^Dimension DESC-SIZE (Dimension. 115 64))
(def ^{:private true} ^long RETRY-LIMIT 5)

(defn- change-cursor [^Component c url]
  (let [csr (.getCursor c)]
    (doto c
      (.addMouseListener
       (proxy [MouseListener][]
	 (mouseEntered  [^MouseEvent e] (.setCursor ^Component (.getSource e) LINK-CURSOR))
	 (mouseExited   [^MouseEvent e] (.setCursor ^Component (.getSource e) csr))
	 (mousePressed  [^MouseEvent e] (p/open-url :alert url))
	 (mouseClicked  [^MouseEvent e])
         (mouseReleased [^MouseEvent e]))))))

(defn alert-panel [^nico.pgm.Pgm pgm ^ImageIcon thumbicn]
  (let [apanel (JPanel.)
        tpanel (JPanel.)
        dpanel (JPanel.)
        owner (su/ifstr (:owner_name pgm) "")
        comm_name (su/ifstr (:comm_name pgm) (:comm_id pgm))
        olabel (JLabel. (format " %s (%s)" owner comm_name))
        time (JLabel. (format "%s  （%d分前に開始）"
                              (if (:member_only pgm) "※コミュ限" "")
                              (tu/minute (tu/interval (:pubdate pgm) (tu/now)))))]
    (let [title (JLabel. ^String (su/ifstr (:title pgm) (name (:id pgm))))
          cbtn (JButton. ^ImageIcon CLOSE-ICON), layout (SpringLayout.)
          close-listener (proxy [ActionListener][]
                             (actionPerformed [e]
                               (let [p (.getParent apanel)]
                                 (when (and p (instance? java.awt.Window p))
                                   (log/debug (format "p is Window: %s" (pr-str p)))
                                   (.dispatchEvent p (WindowEvent. p WindowEvent/WINDOW_CLOSING))
                                   (.removeActionListener cbtn this)))))]
      (doto title (.setFont uu/DEFAULT-FONT))
      (doto layout
        (.putConstraint SpringLayout/NORTH title 0 SpringLayout/NORTH tpanel)
        (.putConstraint SpringLayout/SOUTH title 0 SpringLayout/SOUTH tpanel)
        (.putConstraint SpringLayout/NORTH cbtn 0 SpringLayout/NORTH tpanel)
        (.putConstraint SpringLayout/SOUTH cbtn 0 SpringLayout/SOUTH tpanel)
        (.putConstraint SpringLayout/WEST title 0 SpringLayout/WEST tpanel)
        (.putConstraint SpringLayout/EAST title 0 SpringLayout/WEST cbtn)
        (.putConstraint SpringLayout/EAST cbtn 0 SpringLayout/EAST tpanel))
      (doto cbtn
        (.addActionListener close-listener)
        (.setPreferredSize CLOSE-ICON-SIZE))
      (when (:member_only pgm) (.setBackground cbtn MONLY-BGCOLOR))
      (doto tpanel
        (.setPreferredSize TITLE-PANEL-SIZE)
        (.setLayout layout) (.add title) (.add cbtn)))
    (let [thumbnail (JLabel. thumbicn),
          ^Component desc (uu/mlabel (su/ifstr (:desc pgm) "") DESC-SIZE)
          layout (GridBagLayout.), c (GridBagConstraints.)]
      (letfn [(set-con!
                [^GridBagLayout lt component x y top left bottom right]
                (set! (.gridx c) x) (set! (.gridy c) y) (set! (.fill c) GridBagConstraints/BOTH)
                (set! (.insets c) (Insets. top left bottom right)) (.setConstraints lt component c))]
        (set-con! layout thumbnail 0 0 0 0 0 0)
        (set-con! layout desc 1 0 0 5 0 0))
      (change-cursor thumbnail (:link pgm)) (change-cursor desc (:link pgm))
      (doto ^Container dpanel
            (.setLayout layout) (.add thumbnail) (.add desc)))
    (doto time (.setHorizontalAlignment (JLabel/RIGHT)) (.setFont uu/DEFAULT-FONT))
    (doto olabel (.setFont uu/DEFAULT-FONT))
    (let [layout (SpringLayout.) s 5 ns -5 is 2]
      (doto layout
        (.putConstraint SpringLayout/WEST tpanel s SpringLayout/WEST apanel)
        (.putConstraint SpringLayout/EAST tpanel ns SpringLayout/EAST apanel)
        (.putConstraint SpringLayout/WEST dpanel s SpringLayout/WEST apanel)
        (.putConstraint SpringLayout/EAST dpanel ns SpringLayout/EAST apanel)
        (.putConstraint SpringLayout/WEST olabel s SpringLayout/WEST apanel)
        (.putConstraint SpringLayout/EAST olabel ns SpringLayout/EAST apanel)
        (.putConstraint SpringLayout/WEST time s SpringLayout/WEST apanel)
        (.putConstraint SpringLayout/EAST time ns SpringLayout/EAST apanel)
        (.putConstraint SpringLayout/NORTH tpanel s SpringLayout/NORTH apanel)
        (.putConstraint SpringLayout/NORTH dpanel s SpringLayout/SOUTH tpanel)
        (.putConstraint SpringLayout/NORTH olabel is SpringLayout/SOUTH dpanel)
        (.putConstraint SpringLayout/NORTH time is SpringLayout/SOUTH olabel)
        (.putConstraint SpringLayout/SOUTH time ns SpringLayout/SOUTH apanel))
      (when (:member_only pgm)
        (.setBackground apanel MONLY-BGCOLOR)
        (.setBackground tpanel MONLY-BGCOLOR)
        (.setBackground dpanel MONLY-BGCOLOR))
      (doto apanel
        (.setLayout layout)
        (.add tpanel) (.add dpanel) (.add olabel) (.add time)))))

File: src/main/java/nico/ui/AlertPanel.java
// -*- coding: utf-8-unix -*-
package nico.ui;
import java.awt.AWTEvent;
import java.awt.Color;
import java.awt.Component;
import java.awt.Container;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.RenderingHints;
import java.awt.Window;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.WindowEvent;
import java.awt.image.BufferedImage;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.swing.ImageIcon;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.LineBorder;
import javax.swing.SwingConstants;

import com.github.sgr.slide.MultiLineLabel;
import nico.ui.AlertPanelLayout.Slot;

public class AlertPanel extends JPanel implements MouseListener {
    private static final Logger log = Logger.getLogger(AlertPanel.class.getCanonicalName());
    public static Font FONT_MSG = new Font(Font.SANS_SERIF, Font.PLAIN, 12);
    public static Color FOREGROUND_DEFAULT = Color.BLACK;
    public static Color BACKGROUND_DEFAULT = Color.WHITE;

    public static AlertPanel create(String msg, List<Image> icons) {
	AlertPanel p = new AlertPanel();
	p.setAlertInfo(msg, icons);
	return p;
    }

    /*
     * AlertPanelを開放する。
     * 本クラスの利用者であるdesktop-alertがdisposeを呼ぶため、releaseでなくdisposeとして実装する。
     */
    public void dispose() {
	for (Component c : getComponents()) {
	    if (c instanceof IconLabel) {
		((IconLabel)c).setIcon(null);
	    }
	}
	removeAll();
	for (Image img : _imgs) {
	    img.flush();
	}
	_imgs.clear();

	_imgs = null;
	_msgLabel.dispose();
	_msgLabel = null;
	setLayout(null);
	_layout = null;
	for (MouseListener l : getMouseListeners()) {
	    removeMouseListener(l);
	}
    }

    private MultiLineLabel _msgLabel = null;
    private ArrayList<Image> _imgs = new ArrayList<Image>();
    private AlertPanelLayout _layout = null;

    public void setAlertInfo(String msg, List<Image> icons) {
	for (Component c : getComponents()) {
	    if (c instanceof IconLabel) {
		((IconLabel)c).setIcon(null);
	    }
	}
	removeAll();
	for (Image img : _imgs) {
	    img.flush();
	}
	_imgs.clear();

	_msgLabel.setText(msg);
	add(_msgLabel, Slot.MSG);
	for (Image img : icons) {
	    Image scaledImg = getScaledImageA(img, AlertPanelLayout.ICON_SIZE.width, AlertPanelLayout.ICON_SIZE.height);
	    _imgs.add(scaledImg);
	    add(new IconLabel(new ImageIcon(scaledImg)), Slot.ICON);
	}
	validate();
    }

    private Image getScaledImageA(Image sourceImage, int width, int height) {
	return sourceImage.getScaledInstance(width, height, Image.SCALE_AREA_AVERAGING);
    }

    private Image getScaledImageB(Image sourceImage, int width, int height) {
	BufferedImage scaledImg = new BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB);
	Graphics2D g = scaledImg.createGraphics();
	g.setRenderingHint(RenderingHints.KEY_ALPHA_INTERPOLATION, RenderingHints.VALUE_ALPHA_INTERPOLATION_QUALITY);
	g.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
	g.setRenderingHint(RenderingHints.KEY_COLOR_RENDERING, RenderingHints.VALUE_COLOR_RENDER_QUALITY);
	g.setRenderingHint(RenderingHints.KEY_DITHERING, RenderingHints.VALUE_DITHER_ENABLE);
	g.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
	g.setRenderingHint(RenderingHints.KEY_RENDERING, RenderingHints.VALUE_RENDER_QUALITY);
	g.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BILINEAR);
	g.setRenderingHint(RenderingHints.KEY_FRACTIONALMETRICS, RenderingHints.VALUE_FRACTIONALMETRICS_ON);
	g.setRenderingHint(RenderingHints.KEY_STROKE_CONTROL, RenderingHints.VALUE_STROKE_NORMALIZE);
	g.drawImage(sourceImage, 0, 0, width, height, null);
	log.log(Level.FINE, MessageFormat.format("scaled image (B): ({0}, {1}) -> ({2}, {3})",
						 sourceImage.getWidth(null), sourceImage.getHeight(null),
						 scaledImg.getWidth(null), scaledImg.getHeight(null)));
	g.dispose();
	return scaledImg;
    }

    private AlertPanel() {
	_msgLabel = new MultiLineLabel();
	_msgLabel.setFont(FONT_MSG);
	_layout = new AlertPanelLayout();
	setLayout(_layout);
	add(_msgLabel, Slot.MSG);
	setForegroundColor(FOREGROUND_DEFAULT);
	setBackgroundColor(BACKGROUND_DEFAULT);
	addMouseListener(this);
    }

    public void setForegroundColor(Color color) {
	Color fg = color != null ? color : FOREGROUND_DEFAULT;
	_msgLabel.setForeground(fg);
    }

    public void setBackgroundColor(Color color) {
	Color bg = color != null ? color : BACKGROUND_DEFAULT;
	setBackground(bg);
    }

    private static class PartLabel extends JLabel {
	public void ignoreMouse(boolean b) {
	    if (b) {
		disableEvents(AWTEvent.MOUSE_EVENT_MASK |
			      AWTEvent.MOUSE_MOTION_EVENT_MASK |
			      AWTEvent.MOUSE_WHEEL_EVENT_MASK);
	    } else {
		enableEvents(AWTEvent.MOUSE_EVENT_MASK |
			     AWTEvent.MOUSE_MOTION_EVENT_MASK |
			     AWTEvent.MOUSE_WHEEL_EVENT_MASK);
	    }
	}
    }

    private static class IconLabel extends PartLabel {
	public IconLabel() {
	    setOpaque(false);
	    setHorizontalTextPosition(SwingConstants.CENTER);
	    setHorizontalAlignment(SwingConstants.CENTER);
	    setBorder(new LineBorder(Color.lightGray, 1, false));
	}

	public IconLabel(ImageIcon icon) {
	    this();
	    setIcon(icon);
	}
    }

    // MouseListener
    public void mouseClicked(MouseEvent e) {
	Container c = getParent();
	if ((c != null) && (c instanceof Window)) {
	    c.dispatchEvent(new WindowEvent((Window)c, WindowEvent.WINDOW_CLOSING));
	}
    }

    public void mouseEntered(MouseEvent e) {
    }

    public void mouseExited(MouseEvent e) {
    }

    public void mousePressed(MouseEvent e) {
    }

    public void mouseReleased(MouseEvent e) {
    }
}
